File: foreign/go/contracts/create_user.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package iggcon

import "encoding/binary"

type CreateUser struct {
	Username    string       `json:"username"`
	Password    string       `json:"Password"`
	Status      UserStatus   `json:"Status"`
	Permissions *Permissions `json:"Permissions,omitempty"`
}

func (c *CreateUser) Code() CommandCode {
	return CreateUserCode
}

func (c *CreateUser) MarshalBinary() ([]byte, error) {
	capacity := 4 + len(c.Username) + len(c.Password)
	if c.Permissions != nil {
		capacity += 1 + 4 + c.Permissions.Size()
	}

	bytes := make([]byte, capacity)
	position := 0

	bytes[position] = byte(len(c.Username))
	position += 1
	copy(bytes[position:position+len(c.Username)], []byte(c.Username))
	position += len(c.Username)

	bytes[position] = byte(len(c.Password))
	position += 1
	copy(bytes[position:position+len(c.Password)], []byte(c.Password))
	position += len(c.Password)

	statusByte := byte(0)
	switch c.Status {
	case Active:
		statusByte = byte(1)
	case Inactive:
		statusByte = byte(2)
	}
	bytes[position] = statusByte
	position += 1

	if c.Permissions != nil {
		bytes[position] = byte(1)
		position += 1
		permissionsBytes, err := c.Permissions.MarshalBinary()
		if err != nil {
			return nil, err
		}
		binary.LittleEndian.PutUint32(bytes[position:position+4], uint32(len(permissionsBytes)))
		position += 4
		copy(bytes[position:position+len(permissionsBytes)], permissionsBytes)
	} else {
		bytes[position] = byte(0)
	}

	return bytes, nil
}

File: core/common/src/http/users/create_user.rs
/* Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

use super::defaults::*;
use crate::Permissions;
use crate::UserStatus;
use crate::Validatable;
use crate::error::IggyError;
use secrecy::{ExposeSecret, SecretString};
use serde::{Deserialize, Serialize};

/// `CreateUser` command is used to create a new user.
/// It has additional payload:
/// - `username` - unique name of the user, must be between 3 and 50 characters long.
/// - `password` - password of the user, must be between 3 and 100 characters long.
/// - `status` - status of the user, can be either `active` or `inactive`.
/// - `permissions` - optional permissions of the user. If not provided, user will have no permissions.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CreateUser {
    /// Unique name of the user, must be between 3 and 50 characters long.
    pub username: String,
    /// Password of the user, must be between 3 and 100 characters long.
    #[serde(serialize_with = "crate::utils::serde_secret::serialize_secret")]
    pub password: SecretString,
    /// Status of the user, can be either `active` or `inactive`.
    pub status: UserStatus,
    /// Optional permissions of the user. If not provided, user will have no permissions.
    pub permissions: Option<Permissions>,
}

impl Default for CreateUser {
    fn default() -> Self {
        CreateUser {
            username: "user".to_string(),
            password: SecretString::from("secret"),
            status: UserStatus::Active,
            permissions: None,
        }
    }
}

impl Validatable<IggyError> for CreateUser {
    fn validate(&self) -> Result<(), IggyError> {
        if self.username.is_empty()
            || self.username.len() > MAX_USERNAME_LENGTH
            || self.username.len() < MIN_USERNAME_LENGTH
        {
            return Err(IggyError::InvalidUsername);
        }

        let password = self.password.expose_secret();
        if password.is_empty()
            || password.len() > MAX_PASSWORD_LENGTH
            || password.len() < MIN_PASSWORD_LENGTH
        {
            return Err(IggyError::InvalidPassword);
        }

        Ok(())
    }
}
